src/compiler/utilities.ts

```diff
@@ -7171,30 +7170,70 @@ namespace ts {
         return d1.relatedInformation ? Comparison.LessThan : Comparison.GreaterThan;
     }
 
+    // function compareMessageText(t1: string | DiagnosticMessageChain[], t2: string | DiagnosticMessageChain[]): Comparison {
+    //     if (typeof t1 === 'string' && typeof t2 === 'string') {
+    //         return compareStringsCaseSensitive(t1, t2)
+    //     }
+    //     else if (Array.isArray(t1) && Array.isArray(t2)) {
+    //         if (t1.length < t2.length) {
+    //             return Comparison.LessThan;
+    //         }
+    //         else if (t1.length > t2.length) {
+    //             return Comparison.GreaterThan;
+    //         }
+    //         else {
+    //             for (let i = 0; i < t1.length; i++) {
+    //                 t1[i].messageText
+    //                 const res = cmps(t1[i], t2[i]);
+    //                 if (res) {
+    //                     return res;
+    //                 }
+    //             }
+    //             return Comparison.EqualTo;
+    //         }
+    //     }
+    //     else if (typeof t1 === 'string') {
+    //         return Comparison.LessThan;
+    //     }
+    //     else {
+    //         return Comparison.GreaterThan;
+    //     }
+    // }
+
     function compareMessageText(t1: string | DiagnosticMessageChain, t2: string | DiagnosticMessageChain): Comparison {
-        let text1: string | DiagnosticMessageChain | undefined = t1;
-        let text2: string | DiagnosticMessageChain | undefined = t2;
-        while (text1 && text2) {
-            // We still have both chains.
-            const string1 = isString(text1) ? text1 : text1.messageText;
-            const string2 = isString(text2) ? text2 : text2.messageText;
-
-            const res = compareStringsCaseSensitive(string1, string2);
+        if (typeof t1 === "string" && typeof t2 === "string") {
+            return compareStringsCaseSensitive(t1, t2);
+        }
+        else if (typeof t1 === "string") {
+            return Comparison.LessThan;
+        }
+        else if (typeof t2 === "string") {
+            return Comparison.GreaterThan;
+        }
+        let res = compareStringsCaseSensitive(t1.messageText, t2.messageText);
+        if (res) {
+            return res;
+        }
+        if (!t1.next && !t2.next) {
+            return Comparison.EqualTo;
+        }
+        if (!t1.next) {
+            return Comparison.LessThan;
+        }
+        if (!t2.next) {
+            return Comparison.GreaterThan;
+        }
+        res = compareValues(t1.next.length, t2.next.length);

```

Comment:
Rather than sorting by length first, I think we should should by "effective alphanumeric concatenation" - so _first_ we sort pairwise over the elements of `next` until either both are exhausted or once side is missing one - if one side is missing one, the side with fewer is then less. This ordering has the effect that
```ts
1. - Message qux
    - Message vaz
2. - Message alph
3. - Message wex
```
will be sorted
```ts
2. - Message alph
1. - Message qux
    - Message vaz
3. - Message wex
```
because we'll sort the `qux` message between `wex` and `alph` before sorting by length (which we won't need to do, since everything will be sorted just by looking at the first message).